File: reference/CodeExamples/dotNet/CSharpExamples/Code/concurrency/transactions/DatabaseSupportWithReadWriteLock.cs
using System;
using System.Threading;
using Db4objects.Db4o;

namespace Db4oDoc.Code.Concurrency.Transactions
{
    public class DatabaseSupportWithReadWriteLock : IDisposable
    {
        private readonly ReaderWriterLockSlim transactionLock = new ReaderWriterLockSlim();
        private readonly IObjectContainer database;

        public DatabaseSupportWithReadWriteLock(IObjectContainer database)
        {
            this.database = database;
        }

        // #example: The read transaction method
        public T InWriteTransaction<T>(TransactionFunction<T> transactionClosure)
        {
            transactionLock.EnterWriteLock();
            try
            {
                return transactionClosure(database);
            }
            catch (Exception)
            {
                database.Rollback();
                throw;
            }
            finally
            {
                database.Commit();
                transactionLock.ExitWriteLock();
            }
        }
        // #end example
        // #example: The write transaction method
        public T InReadTransaction<T>(TransactionFunction<T> transactionClosure)
        {
            transactionLock.EnterReadLock();
            try
            {
                return transactionClosure(database);
            }
            catch (Exception)
            {
                database.Rollback();
                throw;
            }
            finally
            {
                database.Commit();
                transactionLock.ExitReadLock();
            }
        }
        // #end example

        public void InWriteTransaction(TransactionAction transactionClosure)
        {
            const int voidReturn = 0;
            InWriteTransaction(container =>
                              {
                                  transactionClosure(container);
                                  return voidReturn;
                              });
        }

        public void InReadTransaction(TransactionAction transactionClosure)
        {
            const int voidReturn = 0;
            InReadTransaction(container =>
            {
                transactionClosure(container);
                return voidReturn;
            });
        }

        public void Dispose()
        {
            database.Dispose();
        }
    }
}

File: reference/CodeExamples/dotNet/VisualBasicExamples/Code/concurrency/transactions/DatabaseSupportWithReadWriteLock.vb
Imports System.Threading
Imports Db4objects.Db4o

Namespace Db4oDoc.Code.Concurrency.Transactions
    Public Class DatabaseSupportWithReadWriteLock
        Implements IDisposable
        Private ReadOnly transactionLock As New ReaderWriterLockSlim()
        Private ReadOnly database As IObjectContainer

        Public Sub New(database As IObjectContainer)
            Me.database = database
        End Sub

        ' #example: The read transaction method
        Public Function InWriteTransaction(Of T)(transactionClosure As TransactionFunction(Of T)) As T
            transactionLock.EnterWriteLock()
            Try
                Return transactionClosure(database)
            Catch generatedExceptionName As Exception
                database.Rollback()
                Throw
            Finally
                database.Commit()
                transactionLock.ExitWriteLock()
            End Try
        End Function
        ' #end example
        ' #example: The write transaction method
        Public Function InReadTransaction(Of T)(transactionClosure As TransactionFunction(Of T)) As T
            transactionLock.EnterReadLock()
            Try
                Return transactionClosure(database)
            Catch generatedExceptionName As Exception
                database.Rollback()
                Throw
            Finally
                database.Commit()
                transactionLock.ExitReadLock()
            End Try
        End Function
        ' #end example

        Public Sub InWriteTransaction(transactionClosure As TransactionAction)
            Const voidReturn As Integer = 0
            InWriteTransaction(Function(container)
                                   transactionClosure(container)
                                   Return voidReturn

                               End Function)
        End Sub

        Public Sub InReadTransaction(transactionClosure As TransactionAction)
            Const voidReturn As Integer = 0
            InReadTransaction(Function(container)
                                  transactionClosure(container)
                                  Return voidReturn

                              End Function)
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
            database.Dispose()
        End Sub
    End Class
End Namespace
